File: velocity/src/main/java/ru/cororo/authserver/velocity/internal/Messages.java
package ru.cororo.authserver.velocity.internal;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.util.Locale;
import java.util.Map;
import java.util.Properties;

/** Proxy command texts in MiniMessage, bundled in English and Russian; the client locale picks one. */
final class Messages {
    private static final Map<String, Properties> LANGUAGES = Map.of("en", load("en"), "ru", load("ru"));

    private Messages() {
    }

    static Component render(Locale locale, String key, String... placeholders) {
        Properties language = LANGUAGES.getOrDefault(locale == null ? "en" : locale.getLanguage(), LANGUAGES.get("en"));
        String template = language.getProperty(key, LANGUAGES.get("en").getProperty(key, key));
        TagResolver.Builder resolver = TagResolver.builder();
        for (int i = 0; i + 1 < placeholders.length; i += 2) {
            resolver.resolver(Placeholder.unparsed(placeholders[i], placeholders[i + 1]));
        }
        return MiniMessage.miniMessage().deserialize(template, resolver.build());
    }

    private static Properties load(String language) {
        try (InputStream stream = Messages.class.getResourceAsStream("/messages_" + language + ".properties")) {
            Properties properties = new Properties();
            properties.load(new InputStreamReader(stream, StandardCharsets.UTF_8));
            return properties;
        } catch (IOException exception) {
            throw new UncheckedIOException(exception);
        }
    }
}

File: server/src/main/kotlin/ru/cororo/authserver/server/config/Messages.kt
package ru.cororo.authserver.server.config

import com.charleskorn.kaml.Yaml
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver
import org.slf4j.LoggerFactory
import java.nio.file.Files
import java.nio.file.Path
import java.util.Locale
import kotlin.io.path.extension
import kotlin.io.path.nameWithoutExtension

/**
 * Player-facing texts per language. Bundled languages are copied into the messages directory on first start
 * so they can be edited; keys missing from an edited file fall back to the bundled text.
 */
class Messages private constructor(private val languages: Map<String, Map<String, String>>, private val defaultLanguage: String) {
    private val miniMessage = MiniMessage.miniMessage()

    /** @param placeholders pairs of placeholder name and value, inserted as plain text */
    fun render(locale: Locale, key: String, vararg placeholders: Pair<String, Any>): Component {
        val template = languages[locale.language]?.get(key) ?: languages[defaultLanguage]?.get(key) ?: languages["en"]?.get(key) ?: key
        val resolver = TagResolver.resolver(placeholders.map { (name, value) -> Placeholder.unparsed(name, value.toString()) })
        return miniMessage.deserialize(template, resolver)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(Messages::class.java)
        private val bundled = listOf("en", "ru")
        private val format = MapSerializer(String.serializer(), String.serializer())

        fun load(directory: Path, defaultLanguage: String): Messages {
            Files.createDirectories(directory)
            val languages = HashMap<String, Map<String, String>>()
            for (language in bundled) {
                val text = checkNotNull(Messages::class.java.getResourceAsStream("/messages/$language.yml")).use {
                    it.readBytes().decodeToString()
                }
                val file = directory.resolve("$language.yml")
                if (Files.notExists(file)) Files.writeString(file, text)
                languages[language] = Yaml.default.decodeFromString(format, text)
            }
            Files.list(directory).use { files ->
                files.filter { it.extension == "yml" }.forEach { file ->
                    runCatching { Yaml.default.decodeFromString(format, Files.readString(file)) }
                        .onSuccess { edited -> languages[file.nameWithoutExtension] = languages[file.nameWithoutExtension].orEmpty() + edited }
                        .onFailure { logger.error("Could not read messages from {}: {}", file, it.message) }
                }
            }
            return Messages(languages, defaultLanguage)
        }
    }
}
